src/ddsrt/src/string.c

```diff
@@ -162,3 +162,21 @@ ddsrt_strdup(
 
   return ddsrt_memdup(str, strlen(str) + 1);
 }
+
+char *
+ddsrt_strndup(const char *str, size_t len)
+{
+  char *s;
+  size_t n = 0;
+
+  assert(str != NULL);
+
+  while (str[n] != '\0' && n < len) { n++; }
```

Comment:
You could replace this line with a n =strnlen(str, len); which might be better optimized than your 1-character at a time algorithm.

Reply:
I agree, but since it's an optimization, I think it can be tackled later.